File: BFS/Binary Tree Zigzag Level Order Traversal/Binary Tree Zigzag Level Order Traversal.java

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
import java.util.*;
import java.util.Collections;
class Solution {
    public List<List<Integer>> zigzagLevelOrder(TreeNode root) {
        // Declare a Result List
        List<List<Integer>> ans=new ArrayList<>();
        int level = 0;
        if(root==null)
            return ans;
   // Declare a Queue List
        Queue<TreeNode> q=new LinkedList<>();
   // Add Root element      
        q.offer(root);
   // Loop While queue is Not Empty
        while(!q.isEmpty())
        {
           int n = q.size();
            level++;
          List<Integer> list=new ArrayList<>();
            for(int i=0;i < n; i++)
                {
                    if( q.peek().left!=null )
                        q.offer(q.peek().left);
                    if( q.peek().right!=null )
                        q.offer(q.peek().right);
                    list.add(q.poll().val);
                }
             if(level % 2 == 0)
                Collections.reverse(list);
             ans.add(list);
         }
        return ans;
    }
}

File: BFS/Binary Tree Zigzag Level Order Traversal/Binary Tree Zigzag Level Order Traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    queue<TreeNode*> q;
    vector< vector<int>> res;
    vector<vector<int>> zigzagLevelOrder(TreeNode* root) {
        if(root == NULL) return res;
        q.push(root);
        int level = 0;
           while(!q.empty()) {
            int size = q.size();
            level++;
            vector<int> v;
            while(size--) {
                TreeNode* temp = q.front(); q.pop();
                v.push_back(temp->val);
                if(temp->left)
                    q.push(temp->left);
                if(temp->right)
                    q.push(temp->right);
            }
         // when the level is even, just reverse the vector.
            if(level % 2 == 0)
                reverse(v.begin(), v.end());
            res.push_back(v);
        }   
        return res;
    }
};
